src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs

```diff
@@ -0,0 +1,1414 @@
+            var obj = new ClassWithComplexDictionaryProperty
+            {
+                MyDictionary = new Dictionary<string, ClassWith_StrictAttribute> { ["Key"] = new ClassWith_StrictAttribute { Float = 1 } }
+            };
+            Assert.Equal(@"{""MyDictionary"":{""Key"":{""Float"":1}}}", JsonSerializer.Serialize(obj));
+        }
+
+        public class ClassWithComplexDictionaryProperty
+        {
+            [JsonNumberHandling(JsonNumberHandling.AllowReadingFromString)]
+            public Dictionary<string, ClassWith_StrictAttribute> MyDictionary { get; set; }
+        }
+
+        [Fact]
+        public static void TypeAttributeAppliesTo_CustomCollectionElements()
+        {
+            string json = @"[""1""]";
+            MyCustomList obj = JsonSerializer.Deserialize<MyCustomList>(json);
+            Assert.Equal(json, JsonSerializer.Serialize(obj));
+        }
+
+        [JsonNumberHandling(JsonNumberHandling.AllowReadingFromString | JsonNumberHandling.WriteAsString)]
+        public class MyCustomList : List<int> { }
+
+        [Fact]
+        public static void TypeAttributeAppliesTo_CustomCollectionElements_HonoredWhenProperty()
+        {
+            string json = @"{""List"":[""1""]}";
+            ClassWithCustomList obj = JsonSerializer.Deserialize<ClassWithCustomList>(json);
+            Assert.Equal(json, JsonSerializer.Serialize(obj));
+        }
+
+        public class ClassWithCustomList
+        {
+            public MyCustomList List { get; set; }
+        }
+
+        [Fact]
+        public static void TypeAttributeAppliesTo_CustomDictionaryElements()
+        {
+            string json = @"{""Key"":""1""}";
+            MyCustomDictionary obj = JsonSerializer.Deserialize<MyCustomDictionary>(json);
+            Assert.Equal(json, JsonSerializer.Serialize(obj));
+        }
+
+        [JsonNumberHandling(JsonNumberHandling.AllowReadingFromString | JsonNumberHandling.WriteAsString)]
+        public class MyCustomDictionary : Dictionary<string, int> { }
+
+        [Fact]
+        public static void TypeAttributeAppliesTo_CustomDictionaryElements_HonoredWhenProperty()
+        {
+            string json = @"{""Dictionary"":{""Key"":""1""}}";
+            ClassWithCustomDictionary obj = JsonSerializer.Deserialize<ClassWithCustomDictionary>(json);
+            Assert.Equal(json, JsonSerializer.Serialize(obj));
+        }
+
+        public class ClassWithCustomDictionary
+        {
+            public MyCustomDictionary Dictionary { get; set; }
+        }
+
+        [Fact]
+        public static void Attribute_OnType_NotRecursive()
+        {
+            // Recursive behavior would allow a string number.
+            // This is not supported.
+            string json = @"{""NestedClass"":{""MyInt"":""1""}}";
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<AttributeOnFirstLevel>(json));
```

Comment:
I don't understand this example. Can you please explain what you mean by recursive behavior?

Reply:
Recursive behavior in this context is having the number handling for the property apply to other properties in it's type closure. This test verifies that this is not the case.